### PR TITLE
[1/7] Network refactoring - Address

### PR DESCRIPTION
**Pull request by nikkolasg**
This PR shows the first part of the big #499 PR regarding the use of an Address type which contains also the type of the connection.

### CODE REVIEW

File: network/address.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	// PlainTCP represents a vanilla TCP connection
+	PlainTCP ConnType = "tcp"
+	// TLS represents a TLS encrypted connection over TCP
+	TLS = "tls"
+	// PURB represents a PURB encryption connection over TCP
+	PURB = "purb"
+	// Local represents a channel based connection type
+	Local = "local"
+	// UnvalidConnType represents a non valid connection type
+	UnvalidConnType = "wrong"
```
Comment (ineiti) [inline network/address.go]: `InvalidConnType`

File: network/address.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+func (a Address) Valid() bool {
+	vals := strings.Split(string(a), typeAddressSep)
+	if len(vals) != 2 {
+		return false
+	}
+	if connType(vals[0]) == UnvalidConnType {
+		return false
+	}
+
+	if ip, _, e := net.SplitHostPort(vals[1]); e != nil {
```
Comment (ineiti) [inline network/address.go]: don't we have to test the port here, too?

File: network/address.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		return false
+	} else if ip == "localhost" {
+		// localhost is not recognized by net.ParseIP ?
+		return true
+	} else if net.ParseIP(ip) == nil {
```
Comment (ineiti) [inline network/address.go]: I suppose this also works for IPv6?

Comment (Liamsi) [inline network/address.go]: Works for both: https://golang.org/pkg/net/#ParseIP

**Comment by ineiti** [comment]
I like it, looks good.

**Comment by Liamsi** [comment]
> I like it, looks good.

Yep, nice work @nikkolasg

File: app/lib/server/server.go [hunk 4]
```diff
@@ -319,10 +324,10 @@ func entityListToPublics(el *sda.Roster) []abstract.Point {
 	return publics
 }
 
-func isPublicIP(ip string) bool {
+func isPublicIP(ip network.Address) bool {
```
Comment (ineiti) [inline app/lib/server/server.go]: I propose to move this to `network.Address`, as it is useful in general, no? But then we'd also have to add missing blocks - http://www.tutorialspoint.com/ipv4/ipv4_quick_guide.htm

File: network/address.go (hunk 2 above)
Comment (nikkolasg) [inline network/address.go]: Should I test the port being between 0 & 65536 then ?

Comment (ineiti) [inline network/address.go]: The golang-specs also allow for a string in the port (like `http` or `ftp`). Shall we insist on a port-number? I would be OK we do a `0<=port<=65535`. What do you think?

Comment (Liamsi) [inline network/address.go]: I though `SplitHostPort` does return an error if this isn't the case. Apparently, it doesn't: https://play.golang.org/p/4_Cz1Qln-G 
So yes, if `Valid` should only return true if the structure/syntax of the address is guaranteed to be valid, we should validate that.

Comment (ineiti) [inline network/address.go]: But to follow the go-specs we should also allow port-names like `http`. But I prefer to limit port-names to numbers.

Comment (Liamsi) [inline network/address.go]: Hmm, true. I would also prefer to limit port-names to numbers. But I step back a little on this one: it isn't critical (if the port is invalid we would recognize it while establishing the connection) anyways.

**Pull request merged**